### PR TITLE
Arrow batches documentation

### PR DESCRIPTION
**Pull request by sfc-gh-astachowski**
### Description

Added documentation of arrow batches

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -420,6 +420,44 @@ data types. The columns are:
...
+
+	records1, _ := batch.Fetch()
+	records2, _ := batch.WithContext(ctx).Fetch()
+
+will produce the same result in records1 and records2, irrespective of the newly provided ctx.
+
+Fetch():
+Returns the underlying records as *[]arrow.Record. When this function is called, the ArrowBatch checks whether
+the underlying data has already been loaded, and downloads it if not.
+
```
Comment (sfc-gh-pfus) [inline doc.go:460]: There are some parameters missing, worth describing. They are described below, but perhaps we can move them here?
- WithArrowBatchesTimestampOption
- WithHigherPrecision
- WithArrowBatchesUtf8Validation
If they have proper doc around the method we don't have to explain it, but it is worth mentioning they exist.

File: doc.go [hunk 2]
```diff
@@ -420,6 +420,44 @@ data types. The columns are:
 
 Note: SQL NULL values are converted to Golang nil values, and vice-versa.
 
+# Arrow batches
```
Comment (sfc-gh-pfus) [inline doc.go:423]: Such a section already exists, maybe we can merge them?

File: doc.go [hunk 3]
```diff
@@ -420,6 +420,44 @@ data types. The columns are:
...
+
+When working with the arrow columnar format in go driver, ArrowBatch structs are used. These are structs
+corresponding 1:1 to data chunks received from the backend. They allow for access to specific arrow.Record structs.
+In normal circumstances arrow batches are transparent to the user and only act as collections of rows.
+
+An ArrowBatch can exist in a state were the underlying data has not yet been loaded. The data is downloaded and
+translated only as it is required. Translation options are retrieved from a context.Context interface, which is either
+passed from session or set by the user using WithContext(ctx) method.
```
[the remark this answers is not in the document]
Comment (sfc-gh-astachowski) [reply doc.go:431]: Yeah, correct, my bad

File: doc.go [hunk 4]
```diff
@@ -420,6 +420,44 @@ data types. The columns are:
...
+
+The motivation for exposing arrow batches to the user is to allow for simpler and more efficient access to
```
[the remark this answers is not in the document]
Comment (sfc-gh-astachowski) [reply doc.go:433]: I think `allow for` is a lighter term, not suggesting it is required. In the context of making something easier, I would stick to this wording

File: doc.go [hunk 5]
```diff
@@ -420,6 +420,44 @@ data types. The columns are:
...
+irrespective of it’s actual size.
+
+WitchContext(ctx context.Context):
+Sets the context of the ArrowBatch to the one provided. Note that the context will not retroactively apply to data
+that has already been downloaded. For example:
+
+	records1, _ := batch.Fetch()
+	records2, _ := batch.WithContext(ctx).Fetch()
+
+will produce the same result in records1 and records2, irrespective of the newly provided ctx.
```
[the remark this answers is not in the document]
Comment (sfc-gh-astachowski) [reply doc.go:455]: Yes, that's the point I was trying to make here

File: doc.go [hunk 6]
```diff
@@ -648,6 +655,31 @@ You must use `WithArrowBatches` context, similar to the following:
 
 		... // use Arrow records
 
+This returns []*ArrowBatch.
+
+ArrowBatch methods:
```
Comment (sfc-gh-pfus) [inline doc.go:660]: functions?

File: doc.go [hunk 7]
```diff
@@ -648,6 +655,31 @@ You must use `WithArrowBatches` context, similar to the following:
...
+
+GetRowCount():
+Returns the number of rows in the ArrowBatch. Note that this returns 0 if the data has not yet been loaded,
+irrespective of it’s actual size.
+
+WitchContext(ctx context.Context):
```
Comment (sfc-gh-pfus) [inline doc.go:666]: Burn this witch!

File: doc.go [hunk 8]
```diff
@@ -648,6 +655,31 @@ You must use `WithArrowBatches` context, similar to the following:
...
+	records2, _ := batch.WithContext(ctx).Fetch()
+
+will produce the same result in records1 and records2, irrespective of the newly provided ctx. Context worth noting are:
+-WithArrowBatchesTimestampOption
+-WithHigherPrecision
+-WithArrowBatchesUtf8Validation
+described in more detail later.
+
+Fetch():
+Returns the underlying records as *[]arrow.Record. When this function is called, the ArrowBatch checks whether
```
Comment (sfc-gh-pfus) [inline doc.go:680]: Not `[]*arrow.Record`?

**Review by sfc-gh-pfus — APPROVED** [review]

File: doc.go (hunk 8 above)
Comment (sfc-gh-astachowski) [reply doc.go:680]: Nope, that's actually *[]arrow.Record

**Review by sfc-gh-pbulawa — APPROVED** [review]

**Pull request merged**